pandas/io/formats/format.py

```diff
@@ -561,7 +561,19 @@ def __init__(
         self.sparsify = sparsify
 
         self.float_format = float_format
-        self.formatters = formatters if formatters is not None else {}
+        if formatters is not None and (
```

Comment:
Can you use `is_list_like` from `pandas._libs.lib` here instead? I think should help simplify the logic

Reply:
Not sure if it will simplify it because `is_list_like` returns `True ` to both _dictionaries_ and  _lists_, however we only want it to enter in this statement if `formatters ` is a _list_.

Reply:
I think this can be simplified if you just precede this condition with:

```python
if formatters is None:
    formatters = {}
```

And then go through the conditions. Right now there is a lot of duplication of conditions which makes it tougher to reason about

Reply:
Made an adjustment, is it better now?

Reply:
@WillAyd 

"You can use a protocol class with isinstance() ..." "isinstance() also works with the predefined protocols in typing such as Iterable." https://mypy.readthedocs.io/en/latest/protocols.html#using-isinstance-with-protocols

We could therefore use isinstance checks that match the types added to the function signatures.

in this case the logic could be as simple as..

```
        if isinstance(formatters, Sequence) and len(frame.columns) != len(formatters):
            msg = (
                "Formatters length({flen}) should match"
                " DataFrame number of columns({dlen})"
            ).format(flen=len(formatters), dlen=len(frame.columns))
            raise ValueError(msg)
        self.formatters = formatters if formatters is not None else {}
```

thoughts?

Reply:
Looks good for me and it works!

Reply:
we don't use this pattern in the codebase, so don't make these changes in this PR. it's a discussion point going forward.

Reply:
also, although mentioned previously, formatters type is currently `Union[List[Callable], Tuple[Callable, ...], Dict[Union[str, int], Callable]` since the code uses isinstance with list, tuple and dict for flow control.

going forward, formatters type could be as permissive as  `Union[Sequence[Callable], Mapping[Union[str, int], Callable]`

and then the protocol based isinstance checks would be more applicable.